Self references in CustomAccountInterface impls are broken by SDK v22

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version are you using?

`22.0.0`

### What did you do?

This is an issue @dmkozh saw while updating the soroban-examples and he mentioned it to me.

Implement the custom account interface, using a `Self` reference in the `__check_auth` function.

The code worked before sdk 22, and required updating as follows to build:
```diff
 #[contractimpl]
 impl CustomAccountInterface for IncrementContract {
     type Signature = BytesN<192>;
     type Error = AccError;
     #[allow(non_snake_case)]
     fn __check_auth(
         env: Env,
         signature_payload: Hash<32>,
-         agg_sig: Self::Signature,
+         agg_sig: BytesN<192>,
         _auth_contexts: Vec<Context>,
     ) -> Result<(), AccError> {
         // The sdk module containing access to the bls12_381 functions
         let bls = env.crypto().bls12_381();
```
Ref: https://github.com/stellar/soroban-examples/pull/330/files#diff-6b53b8afe379dad726d8f8ac76049b737fd0b9a935d5fe5b0b5c0ac8a2535be9

### What did you expect to see?

The existing code and the `Self::Signature` should be allowed.

### What did you see instead?

```
   Compiling soroban-bls-signature v0.0.0 (/home/runner/work/soroban-examples/soroban-examples/bls_signature)
error[E0223]: ambiguous associated type
  --> src/lib.rs:78:18
   |
78 |         agg_sig: Self::Signature,
   |                  ^^^^^^^^^^^^^^^
   |
help: if there were a trait named `Example` with associated type `Signature` implemented for `IncrementContractArgs`, you could use the fully-qualified path
   |
78 |         agg_sig: <IncrementContractArgs as Example>::Signature,
   |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Ref: https://github.com/stellar/soroban-examples/actions/runs/12189741202/job/34005512450?pr=330

--- a/soroban-sdk-macros/Cargo.toml
+++ b/soroban-sdk-macros/Cargo.toml
@@ -23,7 +23,7 @@ soroban-spec = { workspace = true }
 soroban-spec-rust = { workspace = true }
 soroban-env-common = { workspace = true }
 stellar-xdr = { workspace = true, features = ["curr", "std"] }
-syn = {version="2.0",features=["full"]}
+syn = {version="2.0.77",features=["full"]}
 quote = "1.0"
 proc-macro2 = "1.0"
 itertools = "0.10.5"

--- a/soroban-sdk-macros/src/syn_ext.rs
+++ b/soroban-sdk-macros/src/syn_ext.rs
@@ -1,5 +1,6 @@
+use std::collections::HashMap;
 use proc_macro2::TokenStream;
-use quote::{quote, ToTokens};
+use quote::{format_ident, quote, ToTokens};
 use syn::{
     parse::{Parse, ParseStream},
     punctuated::Punctuated,
@@ -135,7 +136,43 @@ impl Parse for HasFnsItem {
         if lookahead.peek(Token![trait]) {
             input.parse().map(HasFnsItem::Trait)
         } else if lookahead.peek(Token![impl]) {
-            input.parse().map(HasFnsItem::Impl)
+            let mut imp = input.parse::<ItemImpl>()?;
+            // Flatten associated types in functions.
+            let associated = imp
+                .items
+                .iter()
+                .filter_map(|item| match item {
+                    // TODO: Flatten consts to values as well.
+                    ImplItem::Type(i) => Some((i.ident.clone(), i.ty.clone())),
+                    _ => None,
+                })
+                .collect::<HashMap<_, _>>();
+            for item in &mut imp.items {
+                if let ImplItem::Fn(f) = item {
+                    for input in &mut f.sig.inputs {
+                        if let FnArg::Typed(t) = input {
+                            if let Type::Path(TypePath { qself: None, path }) = &mut *t.ty {
+                                let segments = &path.segments;
+                                if segments.len() == 2
+                                    && segments.first()
+                                        == Some(&PathSegment::from(format_ident!("Self")))
+                                {
+                                    if let Some(PathSegment {
+                                        arguments: PathArguments::None,
+                                        ident,
+                                    }) = segments.get(1)
+                                    {
+                                        if let Some(resolved_ty) = associated.get(ident) {
+                                            *t.ty = resolved_ty.clone();
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            Ok(HasFnsItem::Impl(imp))
         } else {
             Err(lookahead.error())
         }

--- a/soroban-sdk-macros/src/syn_ext.rs
+++ b/soroban-sdk-macros/src/syn_ext.rs
@@ -1,11 +1,12 @@
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote, ToTokens};
+use std::collections::HashMap;
 use syn::{
     parse::{Parse, ParseStream},
     punctuated::Punctuated,
     token::Comma,
     AngleBracketedGenericArguments, Attribute, GenericArgument, Path, PathArguments, PathSegment,
-    QSelf, ReturnType, Token, TypePath,
+    ReturnType, Token, TypePath,
 };
 use syn::{
     spanned::Spanned, token::And, Error, FnArg, Ident, ImplItem, ImplItemFn, ItemImpl, ItemTrait,
@@ -216,10 +217,14 @@ fn unpack_result(typ: &Type) -> Option<(Type, Type)> {
 fn flatten_associated_items_in_impl_fns(imp: &mut ItemImpl) {
     // TODO: Flatten associated consts used in functions.
     // Flatten associated types used in functions.
-    let Some((_, trait_, _)) = &imp.trait_ else {
-        return;
-    };
-    let self_ty = &*imp.self_ty;
+    let associated_types = imp
+        .items
+        .iter()
+        .filter_map(|item| match item {
+            ImplItem::Type(i) => Some((i.ident.clone(), i.ty.clone())),
+            _ => None,
+        })
+        .collect::<HashMap<_, _>>();
     let fn_input_types = imp
         .items
         .iter_mut()
@@ -232,29 +237,19 @@ fn flatten_associated_items_in_impl_fns(imp: &mut ItemImpl) {
         })
         .flatten();
     for t in fn_input_types {
-        let span = t.span();
-        if let Type::Path(TypePath {
-            qself: qself @ None,
-            path,
-        }) = t.as_mut()
-        {
-            let segments = path.segments.clone();
-            if segments.first() == Some(&PathSegment::from(format_ident!("Self"))) {
-                *qself = Some(QSelf {
-                    lt_token: Token![<](span),
-                    ty: Box::new(self_ty.clone()),
-                    // The index of the first path segment outside the <>.
-                    // For example, would be 2 for <Type as path::Trait>::AssociatedItem.
-                    position: trait_.segments.len(),
-                    as_token: Some(Token![as](span)),
-                    gt_token: Token![>](span),
-                });
-                *path = trait_.clone();
-                // Add the original path segments after the trait, skipping
-                // the first one which is the Self.
-                for segment in segments.into_iter().skip(1) {
-                    path.segments.push_punct(Token![::](span));
-                    path.segments.push_value(segment);
+        if let Type::Path(TypePath { qself: None, path }) = t.as_mut() {
+            let segments = &path.segments;
+            if segments.len() == 2
+                && segments.first() == Some(&PathSegment::from(format_ident!("Self")))
+            {
+                if let Some(PathSegment {
+                    arguments: PathArguments::None,
+                    ident,
+                }) = segments.get(1)
+                {
+                    if let Some(resolved_ty) = associated_types.get(ident) {
+                        *t.as_mut() = resolved_ty.clone();
+                    }
                 }
             }
         }

--- a/soroban-sdk-macros/src/syn_ext.rs
+++ b/soroban-sdk-macros/src/syn_ext.rs
@@ -225,25 +225,30 @@ fn flatten_associated_items_in_impl_fns(imp: &mut ItemImpl) {
             _ => None,
         })
         .collect::<HashMap<_, _>>();
-    for item in &mut imp.items {
-        if let ImplItem::Fn(f) = item {
-            for input in &mut f.sig.inputs {
-                if let FnArg::Typed(t) = input {
-                    if let Type::Path(TypePath { qself: None, path }) = &mut *t.ty {
-                        let segments = &path.segments;
-                        if segments.len() == 2
-                            && segments.first() == Some(&PathSegment::from(format_ident!("Self")))
-                        {
-                            if let Some(PathSegment {
-                                arguments: PathArguments::None,
-                                ident,
-                            }) = segments.get(1)
-                            {
-                                if let Some(resolved_ty) = associated_types.get(ident) {
-                                    *t.ty = resolved_ty.clone();
-                                }
-                            }
-                        }
+    let fn_input_types = imp
+        .items
+        .iter_mut()
+        .filter_map(|item| match item {
+            ImplItem::Fn(f) => Some(f.sig.inputs.iter_mut().filter_map(|input| match input {
+                FnArg::Typed(t) => Some(&mut t.ty),
+                _ => None,
+            })),
+            _ => None,
+        })
+        .flatten();
+    for t in fn_input_types {
+        if let Type::Path(TypePath { qself: None, path }) = t.as_mut() {
+            let segments = &path.segments;
+            if segments.len() == 2
+                && segments.first() == Some(&PathSegment::from(format_ident!("Self")))
+            {
+                if let Some(PathSegment {
+                    arguments: PathArguments::None,
+                    ident,
+                }) = segments.get(1)
+                {
+                    if let Some(resolved_ty) = associated_types.get(ident) {
+                        *t.as_mut() = resolved_ty.clone();
                     }
                 }
             }

--- a/soroban-sdk-macros/src/syn_ext.rs
+++ b/soroban-sdk-macros/src/syn_ext.rs
@@ -1,12 +1,11 @@
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote, ToTokens};
-use std::collections::HashMap;
 use syn::{
     parse::{Parse, ParseStream},
     punctuated::Punctuated,
     token::Comma,
     AngleBracketedGenericArguments, Attribute, GenericArgument, Path, PathArguments, PathSegment,
-    ReturnType, Token, TypePath,
+    QSelf, ReturnType, Token, TypePath,
 };
 use syn::{
     spanned::Spanned, token::And, Error, FnArg, Ident, ImplItem, ImplItemFn, ItemImpl, ItemTrait,
@@ -217,14 +216,10 @@ fn unpack_result(typ: &Type) -> Option<(Type, Type)> {
 fn flatten_associated_items_in_impl_fns(imp: &mut ItemImpl) {
     // TODO: Flatten associated consts used in functions.
     // Flatten associated types used in functions.
-    let associated_types = imp
-        .items
-        .iter()
-        .filter_map(|item| match item {
-            ImplItem::Type(i) => Some((i.ident.clone(), i.ty.clone())),
-            _ => None,
-        })
-        .collect::<HashMap<_, _>>();
+    let Some((_, trait_, _)) = &imp.trait_ else {
+        return;
+    };
+    let self_ty = &*imp.self_ty;
     let fn_input_types = imp
         .items
         .iter_mut()
@@ -237,19 +232,29 @@ fn flatten_associated_items_in_impl_fns(imp: &mut ItemImpl) {
         })
         .flatten();
     for t in fn_input_types {
-        if let Type::Path(TypePath { qself: None, path }) = t.as_mut() {
-            let segments = &path.segments;
-            if segments.len() == 2
-                && segments.first() == Some(&PathSegment::from(format_ident!("Self")))
-            {
-                if let Some(PathSegment {
-                    arguments: PathArguments::None,
-                    ident,
-                }) = segments.get(1)
-                {
-                    if let Some(resolved_ty) = associated_types.get(ident) {
-                        *t.as_mut() = resolved_ty.clone();
-                    }
+        let span = t.span();
+        if let Type::Path(TypePath {
+            qself: qself @ None,
+            path,
+        }) = t.as_mut()
+        {
+            let segments = path.segments.clone();
+            if segments.first() == Some(&PathSegment::from(format_ident!("Self"))) {
+                *qself = Some(QSelf {
+                    lt_token: Token![<](span),
+                    ty: Box::new(self_ty.clone()),
+                    // The index of the first path segment outside the <>.
+                    // For example, would be 2 for <Type as path::Trait>::AssociatedItem.
+                    position: trait_.segments.len(),
+                    as_token: Some(Token![as](span)),
+                    gt_token: Token![>](span),
+                });
+                *path = trait_.clone();
+                // Add the original path segments after the trait, skipping
+                // the first one which is the Self.
+                for segment in segments.into_iter().skip(1) {
+                    path.segments.push_punct(Token![::](span));
+                    path.segments.push_value(segment);
                 }
             }
         }

--- a/soroban-sdk-macros/src/syn_ext.rs
+++ b/soroban-sdk-macros/src/syn_ext.rs
@@ -1,6 +1,6 @@
-use std::collections::HashMap;
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote, ToTokens};
+use std::collections::HashMap;
 use syn::{
     parse::{Parse, ParseStream},
     punctuated::Punctuated,
@@ -134,44 +134,11 @@ impl Parse for HasFnsItem {
         _ = input.parse::<Token![pub]>();
         let lookahead = input.lookahead1();
         if lookahead.peek(Token![trait]) {
-            input.parse().map(HasFnsItem::Trait)
+            let t = input.parse()?;
+            Ok(HasFnsItem::Trait(t))
         } else if lookahead.peek(Token![impl]) {
-            let mut imp = input.parse::<ItemImpl>()?;
-            // Flatten associated types in functions.
-            let associated = imp
-                .items
-                .iter()
-                .filter_map(|item| match item {
-                    // TODO: Flatten consts to values as well.
-                    ImplItem::Type(i) => Some((i.ident.clone(), i.ty.clone())),
-                    _ => None,
-                })
-                .collect::<HashMap<_, _>>();
-            for item in &mut imp.items {
-                if let ImplItem::Fn(f) = item {
-                    for input in &mut f.sig.inputs {
-                        if let FnArg::Typed(t) = input {
-                            if let Type::Path(TypePath { qself: None, path }) = &mut *t.ty {
-                                let segments = &path.segments;
-                                if segments.len() == 2
-                                    && segments.first()
-                                        == Some(&PathSegment::from(format_ident!("Self")))
-                                {
-                                    if let Some(PathSegment {
-                                        arguments: PathArguments::None,
-                                        ident,
-                                    }) = segments.get(1)
-                                    {
-                                        if let Some(resolved_ty) = associated.get(ident) {
-                                            *t.ty = resolved_ty.clone();
-                                        }
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
-            }
+            let mut imp = input.parse()?;
+            flatten_associated_items_in_impl_fns(&mut imp);
             Ok(HasFnsItem::Impl(imp))
         } else {
             Err(lookahead.error())
@@ -244,5 +211,42 @@ fn unpack_result(typ: &Type) -> Option<(Type, Type)> {
             }
         }
         _ => None,
+    }
+}
+
+fn flatten_associated_items_in_impl_fns(imp: &mut ItemImpl) {
+    // TODO: Flatten associated consts used in functions.
+    // Flatten associated types used in functions.
+    let associated_types = imp
+        .items
+        .iter()
+        .filter_map(|item| match item {
+            ImplItem::Type(i) => Some((i.ident.clone(), i.ty.clone())),
+            _ => None,
+        })
+        .collect::<HashMap<_, _>>();
+    for item in &mut imp.items {
+        if let ImplItem::Fn(f) = item {
+            for input in &mut f.sig.inputs {
+                if let FnArg::Typed(t) = input {
+                    if let Type::Path(TypePath { qself: None, path }) = &mut *t.ty {
+                        let segments = &path.segments;
+                        if segments.len() == 2
+                            && segments.first() == Some(&PathSegment::from(format_ident!("Self")))
+                        {
+                            if let Some(PathSegment {
+                                arguments: PathArguments::None,
+                                ident,
+                            }) = segments.get(1)
+                            {
+                                if let Some(resolved_ty) = associated_types.get(ident) {
+                                    *t.ty = resolved_ty.clone();
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
     }
 }

--- a/soroban-sdk/src/tests.rs
+++ b/soroban-sdk/src/tests.rs
@@ -7,6 +7,7 @@ mod bytes_alloc_vec;
 mod bytes_buffer;
 mod contract_add_i32;
 mod contract_assert;
+mod contract_custom_account_impl;
 mod contract_docs;
 mod contract_duration;
 mod contract_fn;

--- a/soroban-sdk/src/tests/contract_custom_account_impl.rs
+++ b/soroban-sdk/src/tests/contract_custom_account_impl.rs
@@ -1,0 +1,48 @@
+use crate::{self as soroban_sdk, BytesN, IntoVal};
+use soroban_sdk::{
+    auth::{Context, CustomAccountInterface},
+    contract, contracterror, contractimpl,
+    crypto::Hash,
+    Env, Vec,
+};
+
+#[contract]
+pub struct Contract;
+
+#[contracterror]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Error {
+    Fail = 1,
+}
+
+#[contractimpl]
+impl CustomAccountInterface for Contract {
+    type Signature = ();
+    type Error = Error;
+
+    /// Check that the signatures and auth contexts are valid.
+    fn __check_auth(
+        _env: Env,
+        _signature_payload: Hash<32>,
+        _signatures: Self::Signature,
+        _auth_contexts: Vec<Context>,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+#[test]
+fn test_functional() {
+    let e = Env::default();
+    let contract_id = e.register(Contract, ());
+    let payload = BytesN::from_array(&e, &[0; 32]);
+    let signature = ();
+    let auth_context = Vec::new(&e);
+    let result = e.try_invoke_contract_check_auth::<Error>(
+        &contract_id,
+        &payload,
+        signature.into_val(&e),
+        &auth_context,
+    );
+    assert_eq!(result, Ok(()));
+}

--- a/soroban-sdk/test_snapshots/tests/contract_custom_account_impl/test_functional.1.json
+++ b/soroban-sdk/test_snapshots/tests/contract_custom_account_impl/test_functional.1.json
@@ -1,0 +1,76 @@
+{
+  "generators": {
+    "address": 1,
+    "nonce": 0
+  },
+  "auth": [
+    [],
+    []
+  ],
+  "ledger": {
+    "protocol_version": 22,
+    "sequence_number": 0,
+    "timestamp": 0,
+    "network_id": "0000000000000000000000000000000000000000000000000000000000000000",
+    "base_reserve": 0,
+    "min_persistent_entry_ttl": 4096,
+    "min_temp_entry_ttl": 16,
+    "max_entry_ttl": 6312000,
+    "ledger_entries": [
+      [
+        {
+          "contract_data": {
+            "contract": "CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD2KM",
+            "key": "ledger_key_contract_instance",
+            "durability": "persistent"
+          }
+        },
+        [
+          {
+            "last_modified_ledger_seq": 0,
+            "data": {
+              "contract_data": {
+                "ext": "v0",
+                "contract": "CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD2KM",
+                "key": "ledger_key_contract_instance",
+                "durability": "persistent",
+                "val": {
+                  "contract_instance": {
+                    "executable": {
+                      "wasm": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+                    },
+                    "storage": null
+                  }
+                }
+              }
+            },
+            "ext": "v0"
+          },
+          4095
+        ]
+      ],
+      [
+        {
+          "contract_code": {
+            "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+          }
+        },
+        [
+          {
+            "last_modified_ledger_seq": 0,
+            "data": {
+              "contract_code": {
+                "ext": "v0",
+                "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+                "code": ""
+              }
+            },
+            "ext": "v0"
+          },
+          4095
+        ]
+      ]
+    ]
+  },
+  "events": []
+}

--- a/tests/account/src/lib.rs
+++ b/tests/account/src/lib.rs
@@ -21,7 +21,7 @@ impl CustomAccountInterface for Contract {
     fn __check_auth(
         _env: Env,
         _signature_payload: Hash<32>,
-        _signatures: (),
+        _signatures: Self::Signature,
         _auth_contexts: Vec<Context>,
     ) -> Result<(), Error> {
         Ok(())

